pkg/controller/machineset/mcps_controller.go

```diff
@@ -0,0 +1,436 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machineset
+
+import (
+	"context"
+	"fmt"
+	machinev1beta1 "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	"github.com/openshift/cluster-api/pkg/util"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// controllerNameMCPS is the name of this controller
+	controllerNameMCPS    = "MCPS_controller"
+	machineRoleLabel      = "machine.openshift.io/cluster-api-machine-role"
+	machineTypeLable      = "machine.openshift.io/cluster-api-machine-type"
+	masterMachineRoleType = "master"
+	mcpsMasterFinalizer   = "machine.openshift.io/mcps-managed"
+)
+
+var (
+	labelsToCheck = map[string]string{
+		machineRoleLabel: masterMachineRoleType,
+		machineTypeLable: masterMachineRoleType,
+	}
+)
+
+// ReconcileMachineSet reconciles a MachineSet object
+type ReconcileMachineControlPlaneSet struct {
+	client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// newReconciler returns a new reconcile.Reconciler.
+func newMCPSReconciler(mgr manager.Manager) *ReconcileMachineControlPlaneSet {
+	return &ReconcileMachineControlPlaneSet{Client: mgr.GetClient(), scheme: mgr.GetScheme(), recorder: mgr.GetEventRecorderFor(controllerName)}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler.
+func addMCPS(mgr manager.Manager, r reconcile.Reconciler, mapFn handler.ToRequestsFunc) error {
+	// Create a new controller.
+	c, err := controller.New(controllerNameMCPS, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Map Machine changes to MachineSets by machining labels.
+	err = c.Watch(
+		&source.Kind{Type: &machinev1beta1.Machine{}},
+		&handler.EnqueueRequestsFromMapFunc{ToRequests: mapFn},
+	)
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to MachineSet.
+	return c.Watch(
+		&source.Kind{Type: &machinev1beta1.MachineControlPlaneSet{}},
+		&handler.EnqueueRequestForObject{},
+	)
+}
+
+func (r *ReconcileMachineControlPlaneSet) MachineToMCPS(o handler.MapObject) []reconcile.Request {
+	result := []reconcile.Request{}
+	m := &machinev1beta1.Machine{}
+	key := client.ObjectKey{Namespace: o.Meta.GetNamespace(), Name: o.Meta.GetName()}
+	err := r.Client.Get(context.Background(), key, m)
```

Comment:
There is no need for a `Get` here because the `MapObject` already has the object embedded, you can cast it similar to: https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/machinedeployment/machinedeployment_controller.go#L368-L372